### ISSUE TITLE
segmentation fault when calling Tensor::run_backward

### ISSUE THREAD

**Issue by caente** [LaurentMazare/tch-rs#144]
I'm starting to explore tch-rs, and started doing so following this post http://vegapit.com/article/how-to-use-torch-in-rust-with-tch-rs. I expect this is due to my own system, but cannot figured it out. 

This is the code I'm using, which is the first few lines of code of the tutorial:

```rust
fn norm_cdf(x: &Tensor) -> Tensor {
    0.5 * (1.0 + (x / Tensor::from(2.0).sqrt()).erf())
}

fn black76(
    epsilon: &Tensor,
    f: &Tensor,
    k: &Tensor,
    t: &Tensor,
    sigma: &Tensor,
    r: &Tensor,
) -> Tensor {
    let d1 = ((f / k).log() + Tensor::from(0.5) * sigma.pow(2.0) * t) / (sigma * t.sqrt());
    let d2 = &d1 - sigma * t.sqrt();
    epsilon * (-r * t).exp() * (f * norm_cdf(&(epsilon * d1)) - k * norm_cdf(&(epsilon * d2)))
}

fn main() {
    let epsilon = Tensor::from(2.5);
    let f = Tensor::from(2.5);
    let k = Tensor::from(1.5);
    let t = Tensor::from(2.0);
    let sigma = Tensor::from(0.1);
    let r = Tensor::from(5.5);
    let price = &black76(&epsilon, &f, &k, &t, &sigma, &r);
    let price_grad = Tensor::run_backward(&[price], &[&f, &sigma], true, true);
}
```

- When using the latest master with libtorch 1.4(with cxx11 ABI), I get the `segmentation fault (core dumped)` 
- When the crate 0.1.3 with libtorch 1.3(with cxx11 ABI), I get the `segmentation fault (core dumped)` 
- Although using the crate 0.1.3, with libtorch 1.4(with cxx11 ABI) I get this error, that is probably expected?:

```
error: failed to run custom build command for `torch-sys v0.1.3`

Caused by:
  process didn't exit successfully: `/home/caente/Dropbox/workshop/bayesian/tch-incursion/target/debug/build/torch-sys-86b0ebd0a28a4cb6/build-script-build` (exit code: 1)
--- stdout
cargo:rustc-link-search=native=/home/caente/libtorch/lib
TARGET = Some("x86_64-unknown-linux-gnu")
OPT_LEVEL = Some("0")
HOST = Some("x86_64-unknown-linux-gnu")
CXX_x86_64-unknown-linux-gnu = None
CXX_x86_64_unknown_linux_gnu = None
HOST_CXX = None
CXX = None
CXXFLAGS_x86_64-unknown-linux-gnu = None
CXXFLAGS_x86_64_unknown_linux_gnu = None
HOST_CXXFLAGS = None
CXXFLAGS = None
CRATE_CC_NO_DEFAULTS = None
DEBUG = Some("true")
CARGO_CFG_TARGET_FEATURE = Some("fxsr,mmx,sse,sse2")
```

**Comment by LaurentMazare** [comment]
The last error is probably expected indeed as it should fail to build in this case.
As for the other errors, it's a bit hard to tell without more details. Could you run `cargo test` from master using libtorch 1.4 with the cxx11 abi? (this seems to work in CI and on all the linux boxes that I've been trying so far)
Probably worth trying to run it within gdb to see if it points at something specific.

**Comment by caente** [comment]
all test passed... in tch-rs. 
regarding dbg, I've never used, I looked around and it is not clear how to use it with rust... do you know any resource that could help me with that?

**Comment by vegapit** [comment]
> I'm starting to explore tch-rs, and started doing so following this post http://vegapit.com/article/how-to-use-torch-in-rust-with-tch-rs. I expect this is due to my own system, but cannot figured it out.
> 
> This is the code I'm using, which is the first few lines of code of the tutorial:
> 
> ```rust
> fn norm_cdf(x: &Tensor) -> Tensor {
>     0.5 * (1.0 + (x / Tensor::from(2.0).sqrt()).erf())
> }
> 
> fn black76(
>     epsilon: &Tensor,
>     f: &Tensor,
>     k: &Tensor,
>     t: &Tensor,
>     sigma: &Tensor,
>     r: &Tensor,
> ) -> Tensor {
>     let d1 = ((f / k).log() + Tensor::from(0.5) * sigma.pow(2.0) * t) / (sigma * t.sqrt());
>     let d2 = &d1 - sigma * t.sqrt();
>     epsilon * (-r * t).exp() * (f * norm_cdf(&(epsilon * d1)) - k * norm_cdf(&(epsilon * d2)))
> }
> 
> fn main() {
>     let epsilon = Tensor::from(2.5);
>     let f = Tensor::from(2.5);
>     let k = Tensor::from(1.5);
>     let t = Tensor::from(2.0);
>     let sigma = Tensor::from(0.1);
>     let r = Tensor::from(5.5);
>     let price = &black76(&epsilon, &f, &k, &t, &sigma, &r);
>     let price_grad = Tensor::run_backward(&[price], &[&f, &sigma], true, true);
> }
> ```
> 
> * When using the latest master with libtorch 1.4(with cxx11 ABI), I get the `segmentation fault (core dumped)`
> * When the crate 0.1.3 with libtorch 1.3(with cxx11 ABI), I get the `segmentation fault (core dumped)`
> * Although using the crate 0.1.3, with libtorch 1.4(with cxx11 ABI) I get this error, that is probably expected?:
> 
> ```
> error: failed to run custom build command for `torch-sys v0.1.3`
> 
> Caused by:
>   process didn't exit successfully: `/home/caente/Dropbox/workshop/bayesian/tch-incursion/target/debug/build/torch-sys-86b0ebd0a28a4cb6/build-script-build` (exit code: 1)
> --- stdout
> cargo:rustc-link-search=native=/home/caente/libtorch/lib
> TARGET = Some("x86_64-unknown-linux-gnu")
> OPT_LEVEL = Some("0")
> HOST = Some("x86_64-unknown-linux-gnu")
> CXX_x86_64-unknown-linux-gnu = None
> CXX_x86_64_unknown_linux_gnu = None
> HOST_CXX = None
> CXX = None
> CXXFLAGS_x86_64-unknown-linux-gnu = None
> CXXFLAGS_x86_64_unknown_linux_gnu = None
> HOST_CXXFLAGS = None
> CXXFLAGS = None
> CRATE_CC_NO_DEFAULTS = None
> DEBUG = Some("true")
> CARGO_CFG_TARGET_FEATURE = Some("fxsr,mmx,sse,sse2")
> ```

Give me 15 minutes and I will have a look. I did not run the code with the latest version of libtorch.

**Comment by vegapit** [comment]
@caente Ok, if you reclone the [repository](https://github.com/vegapit/tchtut), the examples should work now. The problem was that the saved `Cargo.lock` file in the repository was binding the project to an old version of `torch-sys`. I took it out of source versioning so `Cargo` should now redefine the correct one at compile time.

P.S: If you have any more issues, log them directly in the issues of the example repository. I stumbled across your problem by chance =:]

**Comment by caente** [comment]
I was going to continue at https://github.com/vegapit/tchtut but I don't think this issue belongs there.

It turns out that the problem was that I forgot to add `.set_requires_grad(true)` to the two variables we want the derivative of, once added, everything worked. 

It also works with the crate 0.1.3 and libtorch 1.3 

In other words, the issue seems to be catching that "obvious" mistake,  I could venture a PR to improve this, but I'd need a hint for where to start.

**Comment by LaurentMazare** [comment]
I just pushed some changes that should hopefully turn the segfault into a panic with an error message explaining the `set_requires_grad` bit.
Also just to point out that I would recommend not using `Tensor::run_backward` unless you're willing to compute second order derivatives. The typical way to get some gradient would be more along the following lines (taken from the tests):
```rust
    let mut x = Tensor::from(2.0).set_requires_grad(true);
    let y = &x * &x + &x + 36;
    x.zero_grad();
    y.backward();
    let dy_over_dx = x.grad();
```
